File: app/src/main/java/com/paxus/pay/poslinkui/demo/utils/TaskScheduler.java
package com.paxus.pay.poslinkui.demo.utils;

import android.app.Activity;
import android.os.Bundle;

import com.paxus.pay.poslinkui.demo.entry.task.FinishTask;
import com.paxus.pay.poslinkui.demo.entry.task.ScheduledTask;
import com.paxus.pay.poslinkui.demo.entry.task.TimeoutTask;

import java.lang.reflect.InvocationTargetException;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;

public class TaskScheduler {

//Static Members
    public static final String SCHEDULE = "schedule", PARAM_DELAY = "delay", PARAM_TASK = "taskType", PARAM_INIT_TIME = "initTime";
    public enum TASK {TIMEOUT, FINISH}

    private static Map<TASK, Class> TaskClassMap = new HashMap<TASK, Class>(){{
        put(TASK.TIMEOUT, TimeoutTask.class);
        put(TASK.FINISH, FinishTask.class);
    }};

    //This creates bundle necessary to send scheduling request from fragment to entryactivity
    public static Bundle generateTaskRequestBundle(TASK taskType, long delay){
        return new BundleMaker().addLong(PARAM_DELAY, delay).addString(PARAM_TASK, taskType.name()).addLong(PARAM_INIT_TIME, System.currentTimeMillis()).get();
    }

//Instance
    private Activity activityForContext;
    ScheduledExecutorService scheduledExecutorService;
    private Map<TASK, ScheduledTask> scheduledTaskMap = new HashMap<>();

    public TaskScheduler(Activity activity){
        activityForContext = activity;
        scheduledExecutorService = Executors.newSingleThreadScheduledExecutor();
    }

    public void shutdown(){
        cancelTasks();
        scheduledExecutorService.shutdownNow();
    }

    public void cancelTask(TASK taskType){
        if(scheduledTaskMap.containsKey(taskType.name())){
            scheduledTaskMap.get(taskType.name()).cancel();
            scheduledTaskMap.remove(taskType.name());
        }
    }
    public void cancelTasks() {
        for(Map.Entry<TASK, ScheduledTask> scheduledTaskEntry : scheduledTaskMap.entrySet()){
            scheduledTaskEntry.getValue().cancel();
        }
        scheduledTaskMap.clear();
    }

    public void schedule(TASK taskType, long delay, long initTime){
        if(scheduledTaskMap.containsKey(taskType)){
            if(scheduledTaskMap.get(taskType).getDelay() < delay){
                cancelTask(taskType);
            } else {
                return;
            }
        }
        scheduledTaskMap.put(taskType, scheduleNewTask(activityForContext, taskType, delay, initTime));
    }

    private ScheduledTask scheduleNewTask(Activity activity, TASK taskType, long delay, long initTime){
        ScheduledTask scheduledTask = null;
        try {
            scheduledTask = (ScheduledTask) TaskClassMap.get(taskType).getConstructor(Activity.class).newInstance(activity);
        } catch (IllegalAccessException | InstantiationException | InvocationTargetException | NoSuchMethodException e) {
            Logger.e(e);
        }
        ScheduledFuture<?> future = scheduledExecutorService.schedule(scheduledTask, System.currentTimeMillis() - initTime + delay, TimeUnit.MILLISECONDS);
        scheduledTask.attachFuture(future);
        return scheduledTask;
    }

}

File: app/src/main/java/com/paxus/pay/poslinkui/demo/utils/TaskScheduler.kt
package com.paxus.pay.poslinkui.demo.utils

import android.app.Activity
import android.os.Bundle
import com.paxus.pay.poslinkui.demo.entry.task.FinishTask
import com.paxus.pay.poslinkui.demo.entry.task.ScheduledTask
import com.paxus.pay.poslinkui.demo.entry.task.TimeoutTask
import com.paxus.pay.poslinkui.demo.utils.TaskScheduler.TASK
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch

/**
 * Lifecycle-aware task scheduler for timeout and finish tasks.
 *
 * Uses [lifecycleScope] and coroutines for synchronous cancellation during onDestroy.
 * Supports per-requestor task mapping and replace semantics (new task replaces old for same key).
 *
 * @param activityForContext Activity context for task execution
 * @param scope [LifecycleCoroutineScope] (e.g. activity.lifecycleScope) for cancellation
 */
class TaskScheduler(
    private val activityForContext: Activity?,
    private val scope: CoroutineScope
) {
    enum class TASK {
        TIMEOUT, FINISH
    }

    private val jobMap = mutableMapOf<Pair<String, TASK>, Job>()

    /**
     * Cancels all tasks for the given [requestorId]. Call when a Fragment is destroyed.
     */
    fun cancelForRequestor(requestorId: String) {
        val keysToRemove = jobMap.keys.filter { it.first == requestorId }
        keysToRemove.forEach { key ->
            jobMap[key]?.cancel()
            jobMap.remove(key)
        }
        if (keysToRemove.isNotEmpty()) {
            Logger.d("TaskScheduler", "Cancelled tasks for requestor: $requestorId")
        }
    }

    /**
     * Cancels all scheduled tasks.
     */
    fun cancelTasks() {
        jobMap.values.forEach { it.cancel() }
        jobMap.clear()
        Logger.d("TaskScheduler", "Cancelled all tasks")
    }

    /**
     * Shuts down the scheduler. Call from Activity.onDestroy.
     */
    fun shutdown() {
        cancelTasks()
    }

    /**
     * Schedules a task (Activity-level). Uses "activity" as requestorId.
     *
     * @return The scheduled [Job] handle for grouping via [TaskGroup.addToGroup], or null if run immediately
     */
    fun schedule(taskType: TASK, delayMs: Long, initTime: Long): Job? {
        return schedule("activity", taskType, delayMs, initTime)
    }

    /**
     * Schedules a task. If a task for the same (requestorId, taskType) exists, it is replaced.
     *
     * @param requestorId Identifier for the requester (e.g. Fragment)
     * @param taskType TIMEOUT or FINISH
     * @param delayMs Delay in milliseconds
     * @param initTime Initial timestamp for delay calculation
     * @return The scheduled [Job] handle for grouping via [TaskGroup.addToGroup], or null if run immediately
     */
    fun schedule(requestorId: String, taskType: TASK, delayMs: Long, initTime: Long): Job? {
        val key = Pair(requestorId, taskType)
        jobMap[key]?.cancel()
        jobMap.remove(key)

        val remainingDelay = (initTime + delayMs) - System.currentTimeMillis()
        if (remainingDelay <= 0) {
            runTaskImmediately(taskType)
            return null
        }

        val job = scope.launch {
            delay(remainingDelay)
            if (isActive) {
                jobMap.remove(key)
                runTaskImmediately(taskType)
            }
        }
        jobMap[key] = job
        return job
    }

    private fun runTaskImmediately(taskType: TASK) {
        val task: ScheduledTask = when (taskType) {
            TASK.TIMEOUT -> TimeoutTask(activityForContext)
            TASK.FINISH -> FinishTask(activityForContext)
        }
        task.run()
    }

    companion object {
        const val SCHEDULE: String = "schedule"
        const val PARAM_DELAY: String = "delay"
        const val PARAM_TASK: String = "taskType"
        const val PARAM_INIT_TIME: String = "initTime"
        const val PARAM_REQUESTOR_ID: String = "requestorId"

        /**
         * Creates a bundle for scheduling request. Include [requestorId] for per-Fragment cancellation.
         */
        fun generateTaskRequestBundle(taskType: TASK, delay: Long, requestorId: String): Bundle {
            return BundleMaker()
                .addLong(PARAM_DELAY, delay)
                .addString(PARAM_TASK, taskType.name)
                .addLong(PARAM_INIT_TIME, System.currentTimeMillis())
                .addString(PARAM_REQUESTOR_ID, requestorId)
                .get()
        }

        /** Legacy overload without requestorId; uses "activity" as default. */
        fun generateTaskRequestBundle(taskType: TASK, delay: Long): Bundle {
            return generateTaskRequestBundle(taskType, delay, "activity")
        }
    }
}
